homeassistant/components/qbus/coordinator.py

```diff
@@ -0,0 +1,307 @@
+"""Qbus coordinator."""
+
+from __future__ import annotations
+
+import asyncio
+from dataclasses import dataclass
+from datetime import datetime
+import logging
+from typing import Final, cast
+
+from qbusmqttapi.discovery import QbusDiscovery, QbusMqttDevice
+from qbusmqttapi.factory import QbusMqttMessageFactory, QbusMqttTopicFactory
+
+from homeassistant.components.mqtt import (
+    ReceiveMessage,
+    async_wait_for_mqtt_client,
+    client as mqtt,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import CALLBACK_TYPE, Event, HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.event import async_call_later
+from homeassistant.util.hass_dict import HassKey
+
+from .const import CONF_SERIAL_NUMBER, DOMAIN
+from .entity import QbusEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class QbusRuntimeData:
+    """Runtime data for Qbus integration."""
+
+    coordinator: QbusDataCoordinator
+
+
+type QbusConfigEntry = ConfigEntry[QbusRuntimeData]
+QBUS_KEY: HassKey[QbusConfigCoordinator] = HassKey(DOMAIN)
+
+
+class QbusDataCoordinator:
+    """Qbus data coordinator."""
+
+    _WAIT_TIME = 3
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize Qbus coordinator."""
+        _LOGGER.debug("%s - Initializing coordinator", entry.unique_id)
+
+        self._hass = hass
+        self._entry = entry
+
+        self._message_factory = QbusMqttMessageFactory()
+        self._topic_factory = QbusMqttTopicFactory()
+
+        self._cleanup_callbacks: list[CALLBACK_TYPE] = []
+        self._platform_register: dict[
+            str, tuple[type[QbusEntity], AddEntitiesCallback]
+        ] = {}
+        self._registered_entity_ids: list[str] = []
+        self._device_activated = False
+        self._subscribed_to_device_state = False
+        self._device: QbusMqttDevice | None
+
+        # Clean up when HA stops
+        self._cleanup_callbacks.append(
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.shutdown)
+        )
+
+    def shutdown(self, event: Event | None = None) -> None:
+        """Shutdown Qbus coordinator."""
+        _LOGGER.debug("%s - Shutting down coordinator for entry", self._entry.unique_id)
+
+        while self._cleanup_callbacks:
+            self._cleanup_callbacks.pop()()
+
+        self._cleanup_callbacks = []
+        self._registered_entity_ids = []
+        self._device_activated = False
+        self._subscribed_to_device_state = False
+        self._device = None
+
+    def register_platform(
+        self,
+        qbus_type: str,
+        entity_class: type[QbusEntity],
+        add_entities: AddEntitiesCallback,
+    ) -> None:
+        """Register the platform so adding entities can be postponed to when the Qbus config is received."""
```

Comment:
If you extend from `DataUpdateCoordinator`, you can use its helpers to add new devices, instead of storing the platforms here. See https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/dynamic-devices
If you follow that approach, the coordinator does not need to know about platforms, and each platform will handle what devices it should care about.

Reply:
I'm still working on this one.